bytecomp/matching.ml

```diff
@@ -2827,87 +2859,136 @@ let find_in_pat pred =
   end in
   find_rec
 
-let is_lazy_pat = function
-  | Tpat_lazy _ -> true
-  | Tpat_alias _ | Tpat_variant _ | Tpat_record _
-  | Tpat_tuple _|Tpat_construct _ | Tpat_array _
-  | Tpat_or _ | Tpat_constant _ | Tpat_var _ | Tpat_any
-      -> false
-
-let is_lazy p = find_in_pat is_lazy_pat p
-
-let have_mutable_field p = match p with
+let is_mutable_pat p = match p.pat_desc with
 | Tpat_record (lps,_) ->
     List.exists
       (fun (_,lbl,_) ->
         match lbl.Types.lbl_mut with
         | Mutable -> true
         | Immutable -> false)
       lps
+| Tpat_array _ -> true
 | Tpat_alias _ | Tpat_variant _ | Tpat_lazy _
-| Tpat_tuple _|Tpat_construct _ | Tpat_array _
+| Tpat_tuple _|Tpat_construct _
 | Tpat_or _
 | Tpat_constant _ | Tpat_var _ | Tpat_any
```

Comment:
(The above should of course have read "the above line **isn't** quite right")

Reply:
Although you are right in principle, I'll leave the code as it is.
Problems (_i.e._ segfault) occur when the mutable sub-values change status, from scalar to pointer or the other way. In the case of a mutable string, characters may change to other characters, so nothing is to be feared.

Reply:
Yes, I wasn't worried about getting a segfault but about choosing "an incorrect branch" (which, ultimately, was the reason of the segfault in the initial report).

Reply:
Well, what the "correct" branch is in case the subject value is mutated is a moving
target. But well if insist, let us include your patch. Do you insist ?

Reply:
I agree, and no, I don't want to insist right now / on this particular point.

Reply:
(this particular point = mutable strings, in that context)